### PR TITLE
optimize enumerate with start argument

### PR DESCRIPTION
**Pull request by maxbachmann**

### CODE REVIEW

File: Cython/Compiler/FlowControl.py [hunk 1]
```diff
@@ -1016,7 +1016,11 @@ def mark_forloop_target(self, node):
             # Python strings, etc., while correctly falling back to an
             # object type when the base type cannot be handled.
 
-            self.mark_assignment(target, node.item)
+            self.mark_assignment(target, ExprNodes.IndexNode(
+                node.pos,
+                base=sequence,
+                index=ExprNodes.IntNode(target.pos, value='PY_SSIZE_T_MAX',
+                                        type=PyrexTypes.c_py_ssize_t_type)))
```
Comment (maxbachmann) [inline Cython/Compiler/FlowControl.py:1023]: This is the implementation used in TypeInference.py. I am not quite sure why it was changed in https://github.com/cython/cython/commit/0b547d5cf87eed72039ee727deea6d9539ec1898. However I assume it should be the same in both implementations. This fixes #3787

**Comment by scoder** [comment]
Change looks good to me, thanks. Could you please add a test in `tests/run/for_in_string.pyx`? You can use the `typeof()` function to show the inferred type of the loop variable in your example.

File: tests/run/for_in_string.pyx [hunk 2]
```diff
@@ -224,3 +225,27 @@ def for_pyucs4_in_enumerate_unicode(unicode s):
...
+    ('Py_ssize_t', 'Py_UCS4')
+    2
+    >>> for_infered_type_in_enumerate_unicode(unicode_abc_null)
+    ('Py_ssize_t', 'Py_UCS4')
+    'X'
+    >>> for_infered_type_in_enumerate_unicode(unicode_ABC_null)
+    ('Py_ssize_t', 'Py_UCS4')
+    4
+    """
+    print(typeof(i), typeof(c))
```
Comment (scoder) [inline tests/run/for_in_string.pyx:246]: You might want to add a `from __future__ import print_function` at the top of the file.

File: Cython/Compiler/FlowControl.py (hunk 1 above)
Comment (scoder) [reply Cython/Compiler/FlowControl.py:1023]: We're still in alpha phase and 0b547d5 didn't add a test, so … ;-)

File: Cython/Compiler/FlowControl.py [hunk 3]
```diff
@@ -980,7 +980,7 @@ def mark_forloop_target(self, node):
                 if not entry or entry.is_builtin:
                     if function.name == 'reversed' and len(sequence.args) == 1:
                         sequence = sequence.args[0]
-                    elif function.name == 'enumerate' and len(sequence.args) == 1:
+                    elif function.name == 'enumerate' and len(sequence.args) in {1, 2}:
```
Comment (scoder) [inline Cython/Compiler/FlowControl.py:983]: No need to employ hashing here.
```suggestion
                    elif function.name == 'enumerate' and len(sequence.args) in (1, 2):
```

File: Cython/Compiler/TypeInference.py [hunk 4]
```diff
@@ -111,7 +111,7 @@ def visit_ForInStatNode(self, node):
                 if not entry or entry.is_builtin:
                     if function.name == 'reversed' and len(sequence.args) == 1:
                         sequence = sequence.args[0]
-                    elif function.name == 'enumerate' and len(sequence.args) == 1:
+                    elif function.name == 'enumerate' and len(sequence.args) in {1, 2}:
```
Comment (scoder) [inline Cython/Compiler/TypeInference.py:114]:
```suggestion
                    elif function.name == 'enumerate' and len(sequence.args) in (1, 2):
```

File: tests/run/for_in_string.pyx [hunk 5]
```diff
@@ -224,3 +226,27 @@ def for_pyucs4_in_enumerate_unicode(unicode s):
...
+    'X'
+    >>> for_infered_type_in_enumerate_unicode(unicode_ABC)
+    ('Py_ssize_t', 'Py_UCS4')
+    2
+    >>> for_infered_type_in_enumerate_unicode(unicode_abc_null)
+    ('Py_ssize_t', 'Py_UCS4')
+    'X'
+    >>> for_infered_type_in_enumerate_unicode(unicode_ABC_null)
+    ('Py_ssize_t', 'Py_UCS4')
+    4
```
Comment (scoder) [inline tests/run/for_in_string.pyx:245]:
```suggestion
    >>> for_infered_type_in_enumerate_unicode(unicode_abc)
    Py_ssize_t Py_UCS4
    'X'
    >>> for_infered_type_in_enumerate_unicode(unicode_ABC)
    Py_ssize_t Py_UCS4
    2
    >>> for_infered_type_in_enumerate_unicode(unicode_abc_null)
    Py_ssize_t Py_UCS4
    'X'
    >>> for_infered_type_in_enumerate_unicode(unicode_ABC_null)
    Py_ssize_t Py_UCS4
    4
```

**Comment by scoder** [comment]
Thanks!

**Comment by scoder** [comment]
I'll revert this change until https://github.com/cython/cython/issues/4171 is resolved.

**Pull request merged**